### PR TITLE
support errors

### LINKED ISSUE #41: Imperfect handling of short-circuiting for logical operators
The implementation for logical operators like `or` and `and` do not fully account for their short-circuiting behaviors. For example, the expression `x<1 or 1//x` should return `1` for `x=0`, but pysearch cannot find it because `1//x` is immediately pruned due to a division by zero.

The way to deal with this is to make `Num` an `Option<Num>` and allow errors to be represented as `None` instead of pruning them entirely. Fixing this will likely hurt performance, so we would probably want to implement this as an optional flag in params.rs.

Doing this would also allow `None` outputs, which might be useful if the user wants a certain input to throw an error (termination by error is common in code golf :).

### PR DESCRIPTION
**Pull request by 4atj**
resolves #41

### CODE REVIEW

File: src/operator.rs [hunk 1]
```diff
@@ -119,13 +136,27 @@ pub fn apply_or(l: Num, r: Num) -> Option<Num> {
...
 pub fn apply_or_logical(l: Num, r: Num) -> Option<Num> {
-    Some((l != 0 || r != 0) as Num)
+    Some(
+        (l != 0
+            || if Some(r) == ERROR_VALUE {
+                return ERROR_VALUE;
+            } else {
+                r != 0
+            }) as Num,
+    )
```
Comment (lynn) [inline src/operator.rs]: Does something like this work?
```suggestion
    match (l, r) {
        (0, r) if Some(r) == ERROR_VALUE => ERROR_VALUE,
        (l, r) => Some((l != 0 || r != 0) as Num),
    }
```

File: src/params.rs [hunk 2]
```diff
@@ -105,3 +105,5 @@ pub const UNARY_OPERATORS: &[UnaryOp] = &[
 
 /// Match leaf expressions 1 output at a time to avoid unnecessary precalculations
 pub const MATCH_1BY1: bool = true;
+
+pub const ERROR_VALUE: Option<Num> = None;
```
Comment (lynn) [inline src/params.rs:111]: Let's add a comment explaining what this is.

```suggestion
/// If set, e.g. to `Some(-159236)`, this arbitrary number is chosen to represent errors.
/// That is, pysearch will pretend 1/0 = -159236, and -159236 * 2 = -159236, and so on.
pub const ERROR_VALUE: Option<Num> = None;
```

File: src/main.rs [hunk 3]
```diff
@@ -162,7 +162,7 @@ fn find_binary_operators(
                         .iter()
                         .zip(er.output.iter())
                         .enumerate()
-                        .all(|(i, (&ol, &or))| match (op.apply)(ol, or) {
+                        .all(|(i, (&ol, &or))| match op.apply_(ol, or) {
```
Comment (JayXon) [inline src/main.rs:165]: also need to do this for unary 1by1?

Comment (4atj) [reply src/main.rs:165]: indeed

**Review by JayXon — APPROVED** [review]

**Pull request merged**